File: src/main/java/guru/drinkit/controller/SitemapController.java
package guru.drinkit.controller;


import cz.jiripinkas.jsitemapgenerator.ImageBuilder;
import cz.jiripinkas.jsitemapgenerator.WebPage;
import cz.jiripinkas.jsitemapgenerator.WebPageBuilder;
import cz.jiripinkas.jsitemapgenerator.generator.SitemapGenerator;
import guru.drinkit.domain.Ingredient;
import guru.drinkit.domain.Recipe;
import guru.drinkit.repository.IngredientRepository;
import guru.drinkit.repository.RecipeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

import java.util.List;


/**
 * Created by Crabar on 14.05.2016.
 */
@Controller
@RequestMapping("sitemap")
public class SitemapController {

    @Autowired
    RecipeRepository recipeRepository;

    @Autowired
    IngredientRepository ingredientRepository;

    @RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_XML_VALUE)
    @ResponseBody
    public String getSitemap() {
        SitemapGenerator sitemapGenerator = new SitemapGenerator("https://drinkit.guru", new SitemapGenerator.AdditionalNamespace[]{SitemapGenerator.AdditionalNamespace.IMAGE});
        // const pages
        sitemapGenerator.addPage(new WebPageBuilder().build());
        sitemapGenerator.addPage(new WebPageBuilder().name("bar").build());
        // recipes
        List<Recipe> recipes = recipeRepository.findAll();
        for (Recipe recipe : recipes) {
            if (recipe.getPublished()) {
                WebPage recipePage = new WebPageBuilder().name("recipes/" + recipe.getId().toString()).build();
                recipePage.addImage(new ImageBuilder().loc("https://prod-drunkedguru.rhcloud.com" + recipe.getImageUrl())
                        .license("https://creativecommons.org/licenses/by-nc-nd/4.0/")
                        .build());
                sitemapGenerator.addPage(recipePage);
            }
        }
        // ingredients
        List<Ingredient> ingredients = ingredientRepository.findAll();
        for (Ingredient ingredient : ingredients) {
            WebPage ingredientPage = new WebPageBuilder().name("ingredients/" + ingredient.getId().toString()).build();
            sitemapGenerator.addPage(ingredientPage);
        }

        return sitemapGenerator.constructSitemapString();
    }
}

File: src/main/java/guru/drinkit/controller/SitemapController.kt
package guru.drinkit.controller


import cz.jiripinkas.jsitemapgenerator.ImageBuilder
import cz.jiripinkas.jsitemapgenerator.WebPage
import cz.jiripinkas.jsitemapgenerator.WebPageBuilder
import cz.jiripinkas.jsitemapgenerator.generator.SitemapGenerator
import guru.drinkit.domain.Ingredient
import guru.drinkit.domain.Recipe
import guru.drinkit.repository.IngredientRepository
import guru.drinkit.repository.RecipeRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController


/**
 * Created by Crabar on 14.05.2016.
 */
@RestController
@RequestMapping("sitemap")
open class SitemapController
@Autowired
constructor(private val recipeRepository: RecipeRepository,
            private val ingredientRepository: IngredientRepository) {


    @GetMapping(produces = arrayOf(MediaType.APPLICATION_XML_VALUE))
    open fun getSitemap(): String {
        val sitemapGenerator = SitemapGenerator("https://drinkit.guru", arrayOf(SitemapGenerator.AdditionalNamespace.IMAGE))
        sitemapGenerator.addPage(WebPageBuilder().build())
        sitemapGenerator.addPage(WebPageBuilder().name("bar").build())
        sitemapGenerator.addPages(generateRecipesPages())
        sitemapGenerator.addPages(generateIngredientsPages())
        return sitemapGenerator.constructSitemapString()
    }

    private fun generateRecipesPages() = recipeRepository.findAll()
            .filter(Recipe::published)
            .map { convertRecipe(it) }


    private fun convertRecipe(recipe: Recipe): WebPage {
        val recipePage = WebPageBuilder().name("recipes/" + recipe.id.toString()).build()
        recipePage.addImage(ImageBuilder().loc("https://prod-drunkedguru.rhcloud.com" + recipe.imageUrl)
                .license("https://creativecommons.org/licenses/by-nc-nd/4.0/")
                .build())
        return recipePage
    }

    private fun generateIngredientsPages() = ingredientRepository.findAll()
            .map(Ingredient::id)
            .map { WebPageBuilder().name("ingredients/" + it.toString()).build() }

}
